### PR TITLE
Allow Configuration File to Load EMC Values

### PR DESCRIPTION
**Pull request by alycecil**
Allows config file overrides and new defines for items such that the config file always wins.

Also added unit tests to the gradle to be able to test the changes without looping loading minecraft for this. As such I changed a few of the minecraft jar classes out to ones that can be loaded by the gradle orchestration. Ex I replaced Pair.

No idea if this is something you want but I wanted to integrate a couple other mods on one of my servers and figured I'd offer the config loader changes back to you. 

Example config
Integrating croptopia 3.0.9, manure 3.6, and farmingforblockheads 21.0.4
```
emc_on_hud=false # Display current EMC on HUD (top left corner) [default: false]
private_emc=false # Should each player have their own EMC storage? [default: false]
creative_items=true # Should creative items have EMC? [default: false]
difficulty=easy # easy | normal | hard - Changes crafting recipe for Dissolver block. [default: hard]
mode=default # default | skyblock - Changes some EMC values. [default: default]

# manure
emc:manure:manure=1

emc:minecraft:spawner=4200
emc:minecraft:trial_spawner=4200

# spawner eggs cost 4200 like they should for trader
emc:minecraft:allay_spawn_egg=4200
emc:minecraft:axolotl_spawn_egg=4200
emc:minecraft:bat_spawn_egg=4200
emc:minecraft:bee_spawn_egg=4200
emc:minecraft:blaze_spawn_egg=4200
emc:minecraft:camel_spawn_egg=4200
emc:minecraft:cat_spawn_egg=4200
emc:minecraft:cave_spider_spawn_egg=4200
emc:minecraft:chicken_spawn_egg=4200
emc:minecraft:cod_spawn_egg=4200
emc:minecraft:cow_spawn_egg=4200
emc:minecraft:creeper_spawn_egg=4200
emc:minecraft:dolphin_spawn_egg=4200
emc:minecraft:donkey_spawn_egg=4200
emc:minecraft:drowned_spawn_egg=4200
emc:minecraft:elder_guardian_spawn_egg=4200
emc:minecraft:enderman_spawn_egg=4200
emc:minecraft:endermite_spawn_egg=4200
emc:minecraft:evoker_spawn_egg=4200
emc:minecraft:fox_spawn_egg=4200
emc:minecraft:frog_spawn_egg=4200
emc:minecraft:ghast_spawn_egg=4200
emc:minecraft:glow_squid_spawn_egg=4200
emc:minecraft:goat_spawn_egg=4200
emc:minecraft:guardian_spawn_egg=4200
emc:minecraft:hoglin_spawn_egg=4200
emc:minecraft:horse_spawn_egg=4200
emc:minecraft:husk_spawn_egg=4200
emc:minecraft:iron_golem_spawn_egg=4200
emc:minecraft:llama_spawn_egg=4200
emc:minecraft:magma_cube_spawn_egg=4200
emc:minecraft:mooshroom_spawn_egg=4200
emc:minecraft:mule_spawn_egg=4200
emc:minecraft:ocelot_spawn_egg=4200
emc:minecraft:panda_spawn_egg=4200
emc:minecraft:parrot_spawn_egg=4200
emc:minecraft:phantom_spawn_egg=4200
emc:minecraft:pig_spawn_egg=4200
emc:minecraft:piglin_spawn_egg=4200
emc:minecraft:piglin_brute_spawn_egg=4200
emc:minecraft:pillager_spawn_egg=4200
emc:minecraft:polar_bear_spawn_egg=4200
emc:minecraft:pufferfish_spawn_egg=4200
emc:minecraft:rabbit_spawn_egg=4200
emc:minecraft:ravager_spawn_egg=4200
emc:minecraft:salmon_spawn_egg=4200
emc:minecraft:sheep_spawn_egg=4200
emc:minecraft:shulker_spawn_egg=4200
emc:minecraft:silverfish_spawn_egg=4200
emc:minecraft:skeleton_spawn_egg=4200
emc:minecraft:skeleton_horse_spawn_egg=4200
emc:minecraft:slime_spawn_egg=4200
emc:minecraft:sniffer_spawn_egg=4200
emc:minecraft:snow_golem_spawn_egg=4200
emc:minecraft:spider_spawn_egg=4200
emc:minecraft:squid_spawn_egg=4200
emc:minecraft:stray_spawn_egg=4200
emc:minecraft:strider_spawn_egg=4200
emc:minecraft:tadpole_spawn_egg=4200
emc:minecraft:trader_llama_spawn_egg=4200
emc:minecraft:tropical_fish_spawn_egg=4200
emc:minecraft:turtle_spawn_egg=4200
emc:minecraft:vex_spawn_egg=4200
emc:minecraft:villager_spawn_egg=4200
emc:minecraft:vindicator_spawn_egg=4200
emc:minecraft:wandering_trader_spawn_egg=4200
emc:minecraft:warden_spawn_egg=4200
emc:minecraft:witch_spawn_egg=4200
emc:minecraft:wither_skeleton_spawn_egg=4200
emc:minecraft:wolf_spawn_egg=4200
emc:minecraft:zoglin_spawn_egg=4200
emc:minecraft:zombie_spawn_egg=4200
emc:minecraft:zombie_horse_spawn_egg=4200
emc:minecraft:zombie_villager_spawn_egg=4200
emc:minecraft:zombified_piglin_spawn_egg=4200


#croptopia
emc:croptopia:coffee_beans=20

emc:croptopia:calamari=45
emc:croptopia:shrimp=45
emc:croptopia:glowing_calamari=45
emc:croptopia:oyster=45
emc:croptopia:clam=45
emc:croptopia:roe=45
emc:croptopia:sea_lettuce=45

emc:croptopia:artichoke_seed=2
emc:croptopia:asparagus_seed=2
emc:croptopia:barley_seed=2
emc:croptopia:basil_seed=2
emc:croptopia:bellpepper_seed=2
emc:croptopia:blackbean_seed=2
emc:croptopia:blackberry_seed=2
emc:croptopia:blueberry_seed=2
emc:croptopia:broccoli_seed=2
emc:croptopia:cabbage_seed=2
emc:croptopia:cantaloupe_seed=2
emc:croptopia:cauliflower_seed=2
emc:croptopia:celery_seed=2
emc:croptopia:chile_pepper_seed=2
emc:croptopia:coffee_seed=2
emc:croptopia:corn_seed=2
emc:croptopia:cranberry_seed=2
emc:croptopia:cucumber_seed=2
emc:croptopia:currant_seed=2
emc:croptopia:eggplant_seed=2
emc:croptopia:elderberry_seed=2
emc:croptopia:garlic_seed=2
emc:croptopia:ginger_seed=2
emc:croptopia:grape_seed=2
emc:croptopia:greenbean_seed=2
emc:croptopia:greenonion_seed=2
emc:croptopia:honeydew_seed=2
emc:croptopia:hops_seed=2
emc:croptopia:kale_seed=2
emc:croptopia:kiwi_seed=2
emc:croptopia:leek_seed=2
emc:croptopia:lettuce_seed=2
emc:croptopia:mustard_seed=2
emc:croptopia:oat_seed=2
emc:croptopia:olive_seed=2
emc:croptopia:onion_seed=2
emc:croptopia:peanut_seed=2
emc:croptopia:pepper_seed=2
emc:croptopia:pineapple_seed=2
emc:croptopia:radish_seed=2
emc:croptopia:raspberry_seed=2
emc:croptopia:rhubarb_seed=2
emc:croptopia:rice_seed=2
emc:croptopia:rutabaga_seed=2
emc:croptopia:saguaro_seed=2
emc:croptopia:soybean_seed=2
emc:croptopia:spinach_seed=2
emc:croptopia:squash_seed=2
emc:croptopia:strawberry_seed=2
emc:croptopia:sweetpotato_seed=2
emc:croptopia:tea_seed=2
emc:croptopia:tomatillo_seed=2
emc:croptopia:tomato_seed=2
emc:croptopia:turmeric_seed=2
emc:croptopia:turnip_seed=2
emc:croptopia:vanilla_seeds=2
emc:croptopia:yam_seed=2
emc:croptopia:zucchini_seed=2

emc:croptopia:artichoke=20
emc:croptopia:asparagus=20
emc:croptopia:barley=20
emc:croptopia:basil=20
emc:croptopia:bellpepper=20
emc:croptopia:blackbean=20
emc:croptopia:blackberry=20
emc:croptopia:blueberry=20
emc:croptopia:broccoli=20
emc:croptopia:cabbage=20
emc:croptopia:cantaloupe=20
emc:croptopia:cauliflower=20
emc:croptopia:celery=20
emc:croptopia:chile_pepper=20
emc:croptopia:coffee=20
emc:croptopia:corn=20
emc:croptopia:cranberry=20
emc:croptopia:cucumber=20
emc:croptopia:currant=20
emc:croptopia:eggplant=20
emc:croptopia:elderberry=20
emc:croptopia:garlic=20
emc:croptopia:ginger=20
emc:croptopia:grape=20
emc:croptopia:greenbean=20
emc:croptopia:greenonion=20
emc:croptopia:honeydew=20
emc:croptopia:hops=20
emc:croptopia:kale=20
emc:croptopia:kiwi=20
emc:croptopia:leek=20
emc:croptopia:lettuce=20
emc:croptopia:mustard=20
emc:croptopia:oat=20
emc:croptopia:olive=20
emc:croptopia:onion=20
emc:croptopia:peanut=20
emc:croptopia:pepper=20
emc:croptopia:pineapple=20
emc:croptopia:radish=20
emc:croptopia:raspberry=20
emc:croptopia:rhubarb=20
emc:croptopia:rice=20
emc:croptopia:rutabaga=20
emc:croptopia:saguaro=20
emc:croptopia:soybean=20
emc:croptopia:spinach=20
emc:croptopia:squash=20
emc:croptopia:strawberry=20
emc:croptopia:sweetpotato=20
emc:croptopia:tea=20
emc:croptopia:tomatillo=20
emc:croptopia:tomato=20
emc:croptopia:turmeric=20
emc:croptopia:turnip=20
emc:croptopia:vanilla=20
emc:croptopia:yam=20
emc:croptopia:zucchini=20

emc:croptopia:salt=1

emc:croptopia:banana=20
emc:croptopia:apple=20
emc:croptopia:orange=20
emc:croptopia:persimmon=20
emc:croptopia:plum=20
emc:croptopia:cherry=20
emc:croptopia:lemon=20
emc:croptopia:grapefruit=20
emc:croptopia:kumquat=20
emc:croptopia:peach=20
emc:croptopia:coconut=20
emc:croptopia:nutmeg=20
emc:croptopia:fig=20
emc:croptopia:nectarine=20
emc:croptopia:mango=20
emc:croptopia:dragonfruit=20
emc:croptopia:starfruit=20
emc:croptopia:avocado=20
emc:croptopia:apricot=20
emc:croptopia:pear=20
emc:croptopia:lime=20
emc:croptopia:date=20
emc:croptopia:almond=20
emc:croptopia:cashew=20
emc:croptopia:pecan=20
emc:croptopia:walnut=20
emc:croptopia:cinnamon=20

emc:croptopia:cinnamon_sapling=5
```

### CODE REVIEW

**Comment by alycecil** [comment]
Ex spawn egg

4200×3=12600

12600+1863= 14463

Rather than the default of 100k in code in the init block

image
image

**Comment by alycecil** [comment]
Example with croptopia sees values being defined as well. 

image

image

File: src/main/java/net/vassbo/vanillaemc/data/EMCValues.java [hunk 1]
```diff
@@ -65,230 +70,230 @@ public static void init() {
...
-        EMC_VALUES.put("minecraft:pitcher_pod", GRASS);
-        EMC_VALUES.put("minecraft:torchflower_seeds", GRASS);
-        EMC_VALUES.put("minecraft:honeycomb", LEAVES * 10); // this should not be more than 3, because of copper waxing: ((({copper_block_emc}[[+3]])*3+{redstone+blaze_rod})/4)[[+3]]
-        EMC_VALUES.put("minecraft:honey_bottle", LEAVES * 15); // sugar cane * 3
-        EMC_VALUES.put("minecraft:bee_nest", 174); // same as beehive
-        EMC_VALUES.put("minecraft:goat_horn", 1200);
-        EMC_VALUES.put("minecraft:frogspawn", 60);
-        EMC_VALUES.put("minecraft:turtle_egg", 70);
-        EMC_VALUES.put("minecraft:sniffer_egg", 80);
+        setEMCUnchecked("minecraft:cobblestone", COBBLESTONE);
```
Comment (alycecil) [inline src/main/java/net/vassbo/vanillaemc/data/EMCValues.java:73]: these arent strickly needed as this is the initial load, but still do the refactoring just to make it obvious that this is intentional that these puts get direct access to the underlying datastore.

File: src/main/java/net/vassbo/vanillaemc/data/EMCValues.java [hunk 2]
```diff
@@ -422,31 +449,108 @@ public static void init() {
...
+        }
+        for (Map.Entry<String, Integer> entry : NEW_EMC_VALUES.entrySet()) {
+            setEMC(entry.getKey(), entry.getValue());
+        }
+    }
+
     private static boolean tags_loaded = false;
+
     public static void tagsLoaded(HashMap<String, Integer> NEW_EMC_VALUES) {
-        EMC_VALUES.putAll(NEW_EMC_VALUES);
```
Comment (alycecil) [inline src/main/java/net/vassbo/vanillaemc/data/EMCValues.java:427]: this is where the actually required refactoring to use checked methods starts

**Comment by vassbo** [comment]
Nice! Thanks again! Looks good. 😃

**Pull request merged**